File: coding/Leetcode/Dynamic_Programming/task_121.java
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/?envType=study-plan&id=dynamic-programming-i
package DP;

import java.util.Scanner;

public class task_121 {
    public static int maxProfit(int[] prices) {
        int n = prices.length;
        int global_min = prices[0];
        int profit = 0;
        for (int i = 1; i < n; ++i) {
            int temp_profit = prices[i] - global_min;
            if (temp_profit > profit) {
                profit = temp_profit;
            }
            else if (prices[i] < global_min){
                global_min = prices[i];
            }
        }
        return profit;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++)
            nums[i] = sc.nextInt();
        System.out.println(maxProfit(nums));
    }

}

File: coding/Leetcode/Dynamic_Programming/task_121.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Created by omega515 on 28.02.2025.
//
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);

        int answer = 0;
        int n = prices.size();
        int currMax = prices[0], currMin = prices[0], curr;
        for (int i = 1; i < n; ++i) {
            curr = prices[i];
            if (curr > currMin) {
                answer = max(answer, curr - currMin);
            } else {
                currMin = curr;
            }
            currMax = max(currMax, curr);
        }

        return answer;
    }
};

int main() {
    vector<int> prices;
    int answer;
    Solution solve;

    prices = {7, 1, 5, 3, 6, 4};
    answer = 5;
    assert(answer == solve.maxProfit(prices));

    prices = {7, 6, 4, 3, 1};
    answer = 0;
    assert(answer == solve.maxProfit(prices));

    return 0;
}

